False positive `unused-private-member` if class member is mutated via `cls`

### Steps to reproduce

Given a file `w0238.py`:
```python
"""unused-private-member (W0238)"""

class UnusedPrivateMember:
    """unused-private-member (W0238)"""

    __attr_a = None
    __attr_b = 'b'

    @classmethod
    def load_attrs(cls):
        """Load attributes."""
        cls.__attr_a = 'a'

    @property
    def attr_a(self):
        """Get a."""
        return self.__attr_a

    @property
    def attr_b(self):
        """Get b."""
        return self.__attr_b

```

### Current behavior

Result of `pylint w0238.py`:
```
************* Module w0238
w0238.py:12:8: W0238: Unused private member `UnusedPrivateMember.__attr_a` (unused-private-member)
```

### Expected behavior

I would expect no `unused-private-member` warning as `cls` is inferred as the current class. 
However, this example feels error prone under edge cases.

### pylint --version output

Result of `pylint --version` output:

```
pylint 2.9.3
astroid 2.6.2
Python 3.8.3 (default, Jul  3 2020, 09:39:00) 
[GCC 7.5.0]
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -17,6 +17,11 @@ Release date: TBA
 ..
   Put bug fixes that should not wait for a new minor version here
 
+* Fix a false positive for ``unused-private-member`` when mutating a private attribute
+  with ``cls``
+
+  Closes #4657
+
 
 What's New in Pylint 2.9.3?
 ===========================

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -961,9 +961,14 @@ a metaclass class method.",
             if not is_attr_private(assign_attr.attrname):
                 continue
             for attribute in node.nodes_of_class(astroid.Attribute):
-                if (
-                    attribute.attrname == assign_attr.attrname
-                    and attribute.expr.name == assign_attr.expr.name
+                if attribute.attrname == assign_attr.attrname and (
+                    (
+                        # If assigned to cls.attrib, can be accessed by cls/self
+                        assign_attr.expr.name == "cls"
+                        and attribute.expr.name in ["cls", "self"]
+                    )
+                    # If assigned to self.attrib, can only be accessed by self
+                    or (assign_attr.expr.name == attribute.expr.name == "self")
                 ):
                     break
             else:

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -103,3 +103,35 @@ class Klass:
 k = Klass()
 print(k.twentyone)
 print(k.ninetyfive)
+
+
+class FalsePositive4657:
+    """False positivie tests for 4657"""
+    __attr_a = None
+    __attr_b = 'b'
+
+    @classmethod
+    def load_attrs(cls):
+        """Load attributes."""
+        cls.__attr_a = 'a'
+
+    @property
+    def attr_a(self):
+        """Get a."""
+        return self.__attr_a
+
+    @property
+    def attr_b(self):
+        """Get b."""
+        return self.__attr_b
+
+    # Test cases where we assign self.attr, but try to
+    # access cls.attr
+
+    def __init__(self):
+        self.__attr_c = "this is an unused private instance attribute"  # [unused-private-member]
+
+    @property
+    def attr_c(self):
+        """Get c."""
+        return cls.__attr_c  # [undefined-variable]

--- a/tests/functional/u/unused/unused_private_member.txt
+++ b/tests/functional/u/unused/unused_private_member.txt
@@ -1,7 +1,9 @@
-unused-private-member:4:4:AnotherClass.__test:Unused private member `AnotherClass.__test(self)`
-unused-private-member:8:4:HasUnusedInClass:Unused private member `HasUnusedInClass.__my_secret`
-unused-private-member:12:4:HasUnusedInClass.__private_class_method_unused:Unused private member `HasUnusedInClass.__private_class_method_unused(cls)`
-unused-private-member:20:4:HasUnusedInClass.__private_static_method_unused:Unused private member `HasUnusedInClass.__private_static_method_unused()`
-unused-private-member:28:8:HasUnusedInClass.__init__:Unused private member `HasUnusedInClass.__instance_secret`
-unused-private-member:34:4:HasUnusedInClass.__test:Unused private member `HasUnusedInClass.__test(self, x, y, z)`
-unused-private-member:55:4:HasUnusedInClass.__test_recursive:Unused private member `HasUnusedInClass.__test_recursive(self)`
+unused-private-member:4:4:AnotherClass.__test:Unused private member `AnotherClass.__test(self)`:HIGH
+unused-private-member:8:4:HasUnusedInClass:Unused private member `HasUnusedInClass.__my_secret`:HIGH
+unused-private-member:12:4:HasUnusedInClass.__private_class_method_unused:Unused private member `HasUnusedInClass.__private_class_method_unused(cls)`:HIGH
+unused-private-member:20:4:HasUnusedInClass.__private_static_method_unused:Unused private member `HasUnusedInClass.__private_static_method_unused()`:HIGH
+unused-private-member:28:8:HasUnusedInClass.__init__:Unused private member `HasUnusedInClass.__instance_secret`:HIGH
+unused-private-member:34:4:HasUnusedInClass.__test:Unused private member `HasUnusedInClass.__test(self, x, y, z)`:HIGH
+unused-private-member:55:4:HasUnusedInClass.__test_recursive:Unused private member `HasUnusedInClass.__test_recursive(self)`:HIGH
+unused-private-member:132:8:FalsePositive4657.__init__:Unused private member `FalsePositive4657.__attr_c`:HIGH
+undefined-variable:137:15:FalsePositive4657.attr_c:Undefined variable 'cls':HIGH

